articles/zk-snarks-under-the-hood/zk_snarks_under_the_hood.md

```diff
@@ -0,0 +1,139 @@
+
+We have two pairs $(G, k)$ and $(\pi_a, \pi_a')$, and the only way I can know the latter is by using the first pair and multiplying it with A(t).
+
+Check using 3 pairing equations:
+
+$e(\pi_a, k_a) = e(\pi_a', G)$
+
+$e(\pi_b, k_b) = e(\pi_b', G)$
+
+  
+$e(\pi_c, k_c) = e(\pi_c', G)$
+# Generalize
+
+There is one more detail that we need to discuss. Most of the time we don’t just want to prove in the abstract that some solution exists for some specific problem; rather, we want to prove either the correctness of some specific solution (eg. proving that if you take the word “cow” and SHA3 hash it a million times, the final result starts with 0x73064fe5), or that a solution exists if you restrict some of the parameters. For example, in a cryptocurrency instantiation where transaction amounts and account balances are encrypted, you want to prove that you know some decryption key k such that:
+
+1. decrypt(old_balance, k) ≥ decrypt(tx_value, k)
+2. decrypt(old_balance, k) - decrypt(tx_value, k) = decrypt(new_balance, k)
+
+The encrypted old_balance, tx_value and new_balance should be specified publicly, as those are the specific values that we are looking to verify at that particular time; only the decryption key should be hidden. Some slight modifications to the protocol are needed to create a “custom verification key” that corresponds to some specific restriction on the inputs.
+
+# Complexity
+
+Altogether, the verification process is a few elliptic curve multiplications (one for each “public” input variable), and five pairing checks, one of which includes an additional pairing multiplication. The proof contains eight elliptic curve points: a pair of points each for A(t), B(t) and C(t), a point $π_k$ for $b * (A(t) + B(t) + C(t))$, and a point $π_h$ for H(t). Seven of these points are on the $F_p$ curve (32 bytes each, as you can compress the y coordinate to a single bit), and in the Zcash implementation one point ($π_b$) is on the twisted curve in $F_p^2$ (64 bytes), so the total size of the proof is ~288 bytes.
+
+The two computationally hardest parts of creating a proof are:
+
+- Dividing (A * B - C) / Z to get H (algorithms based on the [Fast Fourier transform](https://en.wikipedia.org/wiki/Fast_Fourier_transform) can do this in sub-quadratic time, but it’s still quite computationally intensive)
+- Making the elliptic curve multiplications and additions to create the A(t), B(t), C(t) and H(t) values and their corresponding pairs
+
+The basic reason why creating a proof is so hard is the fact that what was a single binary logic gate in the original computation turns into an operation that must be cryptographically processed through elliptic curve operations if we are making a zero-knowledge proof out of it. This fact, together with the superlinearity of fast Fourier transforms, means that proof creation takes ~20–40 seconds for a Zcash transaction.
+
+# Neglect the Need for Trusted Setup
```

Comment:
Please don't just copy/paste thing

Reply:
My bad, I thought that it was easy enough to understand 😢